### ISSUE TITLE
npm-health: Model Weight and Scoring

### ISSUE THREAD

**Issue by dicortazar** [nexB/scancode.io#2164]
This includes the proposal of the final weight calculation. This very first step may only include the very same 1000 packages. This will bring over fitting to the model

**Comment by canasdiaz** [comment]
This task is blocked by #2166, as we would need to play with data to get metrics and define scores.

**Comment by canasdiaz** [comment]
@dicortazar, we can use something like this to get an initial set of NPM packages for the model https://github.com/tristan-f-r/npm-rank. That list reflects the state of the NPM registry in late 2024.

**Comment by canasdiaz** [comment]
@dicortazar, as we are going to use Census III as a starting point, let's get a medium set of repositories, so we can run GrimioreLab against it.

**Comment by canasdiaz** [comment]
Based on the first iteration using a training set and a testing set, we obtain this equation for the model:

```
=== FINAL MODEL EQUATION ===
z = 0.6484
    +0.9766 * (metric_active_branches)
    +0.3485 * (metric_casual_regular_contributors_rate)
    +0.8720 * (metric_contributor_growth)
    -0.7870 * (metric_developer_categories_core)
    +1.9470 * (metric_elephant_factor)
    +0.0000 * (metric_file_types_binary)
    -0.0951 * (metric_file_types_code)
    +0.0000 * (metric_found_file_adopters)
    -0.2072 * (metric_found_file_license)
    +0.0996 * (metric_pony_factor)
    -0.0701 * (metric_total_commits)
```

So, the probability of project to be healthy would be:

Probability(Healthy) = 1 / (1 + e^(-z))

<img width="374" height="92" alt="Image" src="https://github.com/user-attachments/assets/bb22b704-6e47-45da-b0db-8202cdb4ef2b" />

The immediate step is to drop irrelevant metrics. These are:
 - metric_file_types_binary
 - metric_found_file_adopters

A coefficient of exactly 0.0000 means they add nothing but noise to our dataset.

Do we have to drop more metrics?

I've used Gemini for this and its advice is very interesting. We have to double check this @dicortazar :

> For almost all Unhealthy repositories, almost every single metric is exactly 0. This means metrics like metric_total_commits or metric_active_branches might not actually be measuring "health"—they are likely just measuring whether a repository is active vs. completely archived/dead.

Next steps, tag more projects and run the new set of metrics. We have to make sure the timeframes are the same ones we have in the training data set!

**Comment by canasdiaz** [comment]
@dicortazar let's use this version of the model for the first integration tests. I would like to review more things, but so far I don't find big issues. Let's use this AI generated code snippet to inspire us. 

I've removed the metrics that are not relevant at all. The accuracy is too good, so I'm concerned about overfitting. In any case, so far it looks good enough.  

```
import math

def calculate_project_unhealthiness_score(metrics: dict) -> float:
    """
    Calculates the probability of a repository being 'Unhealthy' based on 
    the pruned logistic regression model metrics.
    
    Parameters:
    metrics (dict): Dictionary containing the project feature names and values.
    
    Returns:
    float: Probability score between 0.0 (Healthy) and 1.0 (Unhealthy).
    """
    # Model Intercept
    z = -0.549873845969752
    
    # Active coefficients mapping (Dropped: found_file_adopters, casual_regular_contributors_rate, file_types_*)
    coefficients = {
        'elephant_factor': -1.635941,
        'coefficient_of_variation': -1.404157,
        'total_contributors': -0.991894,
        'days_since_last_commit': 0.865738,
        'contributor_growth_rate': 0.435875,
        'commits_over_periods_rate': -0.410393,
        'total_commits': -0.330035,
        'message_size_mean': -0.320026,
        'found_file_license': 0.266483
    }
    
    # Calculate the linear combination (log-odds)
    for metric, coef in coefficients.items():
        # Defaults to 0 if a metric is missing, safely neutralizing it
        value = metrics.get(metric, 0.0)
        z += coef * value
        
    # Apply the Sigmoid function to get the final probability
    try:
        probability = 1 / (1 + math.exp(-z))
    except OverflowError:
        # Safeguard against extreme values of z
        probability = 0.0 if z < 0 else 1.0
        
    return probability

# --- Example Usage ---
sample_project = {
    'elephant_factor': 0.4,
    'coefficient_of_variation': 0.8,
    'total_contributors': 15,
    'days_since_last_commit': 5,
    'contributor_growth_rate': 0.12,
    'commits_over_periods_rate': 0.65,
    'total_commits': 450,
    'message_size_mean': 52.4,
    'found_file_license': 1.0,
    
    # The function safely ignores these low-impact metrics if passed
    'found_file_adopters': 1.0, 
    'file_types_binary': 12
}

risk_score = calculate_project_unhealthiness_score(sample_project)
print(f"Project Risk Probability: {risk_score:.2%}")
```

**Comment by canasdiaz** [comment]
This is done, the documentation about how to build the model is available at https://github.com/aboutcode-org/healthycode/tree/main/model/npm

The code is available at https://github.com/aboutcode-org/healthycode/blob/main/src/grimoirelab_metrics/metrics_model.py
